src/Http/Http.Abstractions/src/Api/IFromRouteMetadata.cs

```diff
@@ -0,0 +1,16 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Http.Api
+{
+    /// <summary>
+    /// Interface marking attributes that specify a parameter or property should be bound using route-data from the current request.
+    /// </summary>
+    public interface IFromRouteMetadata
```

Comment:
This doesn't need to change right now but I'm thinking we take the binding source concept from MVC and call it `IBindingSourceMetadata` (or something like that).

```C#
public interface IBindingSourceMetadata
{
    BindingSource BindingSource { get; }
    string Name { get; }
}

public enum BindingSource
{
     Query,
     Form,
     Header,
     Body
}
```

or maybe we need 2 sources (named and unnamed):

```C#
public interface IBindingSourceMetadata
{
    BindingSource BindingSource { get; }
}

public interface INamedBindingSourceMetadata : IBindingSourceMetadata
{
    string Name { get; }
}
```

Spit balling here, but I think we could end up with a single abstraction for binding metadata and we can leave the attributes in MVC.

Reply:
> and call it IBindingSourceMetadata (or something like that)

I was considering just type forwarding the [existing IBindingSourceMetadata](https://github.com/dotnet/aspnetcore/blob/ab9c650a7179010c4a9109a8fc0a8db6ad6d4c3b/src/Mvc/Mvc.Abstractions/src/ModelBinding/BindingSource.cs#L15) which is self-contained, but I know how you feel about having `.Mvc` in the namespace.

Do you want to have `Microsoft.AspNetCore.Mvc.ModelBinding.IBindingSourceMetadata` implement `Microsoft.AspNetCore.Http.ModelBinding.IBindingSourceMetadata` or something like that?

Reply:
I think so. In the short term.

Reply:
Talked offline. We might do this in a subsequent PR, but the kind of stuff I was trying initially where `Mvc.ModelBinding.IBindingSourceMetadata` implements `Http.Metadata.IBindingSourceMetadata` was pretty ugly.

```c#
namespace Microsoft.AspNetCore.Mvc.ModelBinding
{
    /// <summary>
    /// Metadata which specifies the data source for model binding.
    /// </summary>
    public interface IBindingSourceMetadata : Http.Metadata.IBindingSourceMetadata
    {
        /// <summary>
        /// Gets the <see cref="BindingSource"/>. 
        /// </summary>
        /// <remarks>
        /// The <see cref="BindingSource"/> is metadata which can be used to determine which data
        /// sources are valid for model binding of a property or parameter.
        /// </remarks>
        new BindingSource? BindingSource { get; }

        Http.Metadata.BindingSource Http.Metadata.IBindingSourceMetadata.BindingSource
        {
            get => BindingSource?.Id switch
            {
                nameof(BindingSource.Path) => Http.Metadata.BindingSource.Route,
                nameof(BindingSource.Query) => Http.Metadata.BindingSource.Query,
                nameof(BindingSource.Header) => Http.Metadata.BindingSource.Header,
                nameof(BindingSource.Body) => Http.Metadata.BindingSource.Body,
                nameof(BindingSource.Form) => Http.Metadata.BindingSource.Form,
                nameof(BindingSource.Services) => Http.Metadata.BindingSource.Services,
                _ => Http.Metadata.BindingSource.Custom
            };
        }
    }
}
```

Now this doesn't even compile (C# doesn't like interfaces having members with the same name even if explicitly implemented), but you can see the ugliness of having two very similar interfaces.

There are 6 `IFrom*Metadata` attributes so far. I don't think that's too crazy. I do like that the eventual extensibility would be just to create a new interface rather than a second-class `Custom` or `Other` BindingSource.